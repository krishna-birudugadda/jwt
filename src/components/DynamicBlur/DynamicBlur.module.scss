@use 'src/styles/variables';
@use 'src/styles/theme';

.dynamicBlur {
  position: fixed;
  z-index: -1;
}

.image {
  position: absolute;
  top: 0;
  box-sizing: border-box;
  width: 100vw;
  height: 100vh;
  filter: blur(30px) brightness(60%);
}
