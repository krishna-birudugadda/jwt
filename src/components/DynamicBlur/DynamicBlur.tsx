import React, { memo, useEffect, useMemo, useRef, useState } from 'react';

import styles from './DynamicBlur.module.scss';

import { debounce } from '#src/utils/common';
import Fade from '#components/Animation/Fade/Fade';
import Image from '#components/Image/Image';
import type { ImageData } from '#types/playlist';

type Props = {
  image: ImageData;
  transitionTime?: number;
  debounceTime?: number;
};

type ImageCursor = {
  image: ImageData;
  visible: boolean;
  loading: boolean;
  key: string;
};

const DynamicBlur = ({ image, transitionTime = 1, debounceTime = 350 }: Props): JSX.Element => {
  const [images, setImages] = useState<ImageCursor[]>([]);
  const keyRef = useRef(0);
  const updateImage = useMemo(
    () =>
      debounce((image) => {
        setImages((current) => [
          {
            image,
            visible: true,
            loading: true,
            key: `key_${keyRef.current++}`,
          },
          ...current,
        ]);
      }, debounceTime),
    [debounceTime],
  );

  useEffect(() => {
    if (image) updateImage(image);
  }, [updateImage, image]);

  const handleClose = (key: string) => {
    setImages((current) => current.filter((image) => image.key !== key));
  };

  const handleLoad = (key: string) => {
    setImages((current) =>
      current.map((image) => {
        if (image.key === key) {
          return { ...image, loading: false };
        }
        return { ...image, visible: false };
      }),
    );
  };

  return (
    <div className={styles.dynamicBlur}>
      {images.map((cursor) => (
        <Fade
          duration={transitionTime * 1000}
          open={cursor.visible && !cursor.loading}
          delay={cursor.visible ? 100 : 0}
          key={cursor.key}
          onCloseAnimationEnd={() => handleClose(cursor.key)}
          keepMounted
        >
          <Image className={styles.image} image={cursor.image} onLoad={() => handleLoad(cursor.key)} width={1280} />
        </Fade>
      ))}
    </div>
  );
};

export default memo(DynamicBlur);
