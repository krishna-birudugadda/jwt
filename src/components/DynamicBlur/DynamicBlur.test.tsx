import React from 'react';
import { render } from '@testing-library/react';

import DynamicBlur from './DynamicBlur';

describe('<DynamicBlur>', () => {
  test('renders and matches snapshot', () => {
    const { container } = render(<DynamicBlur image={{ image: 'http://image.jpg' }} />);

    expect(container).toMatchSnapshot();
  });
});
