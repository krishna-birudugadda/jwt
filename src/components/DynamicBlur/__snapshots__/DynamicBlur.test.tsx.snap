// Vitest Snapshot v1

exports[`<DynamicBlur> > renders and matches snapshot 1`] = `
<div>
  <div
    class="_dynamicBlur_c7f5f4"
  />
</div>
`;
